CodePush.m

```diff
@@ -554,20 +523,65 @@ - (void)savePendingUpdate:(NSString *)packageHash
+            NSString *previousStatusReportIdentifier = [CodePushStatusReport getPreviousStatusReportIdentifier];
+            if (currentPackageIdentifier) {
+                if (previousStatusReportIdentifier == nil) {
+                    [CodePushStatusReport recordDeploymentStatusReported:currentPackageIdentifier];
+                    resolve(@{
+                              @"package": currentPackage,
+                              @"status": DeploymentSucceeded
+                              });
+                    return;
+                } else if (![previousStatusReportIdentifier isEqualToString:currentPackageIdentifier]) {
+                    [CodePushStatusReport recordDeploymentStatusReported:currentPackageIdentifier];
+                    if ([CodePushStatusReport isStatusReportIdentifierCodePushLabel:previousStatusReportIdentifier]) {
+                        NSString *previousDeploymentKey = [CodePushStatusReport getDeploymentKeyFromStatusReportIdentifier:previousStatusReportIdentifier];
+                        NSString *previousLabel = [CodePushStatusReport getVersionLabelFromStatusReportIdentifier:previousStatusReportIdentifier];
+                        resolve(@{
+                                  @"package": currentPackage,
+                                  @"status": DeploymentSucceeded,
+                                  @"previousDeploymentKey": previousDeploymentKey,
+                                  @"previousLabelOrAppVersion": previousLabel
+                                  });
+                    } else {
+                        // Previous status report was with a binary app version.
+                        resolve(@{
+                                  @"package": currentPackage,
+                                  @"status": DeploymentSucceeded,
+                                  @"previousLabelOrAppVersion": previousStatusReportIdentifier
+                                  });
+                    }
+                    return;
+                }
             }
         }
     } else if (isRunningBinaryVersion || [_bridge.bundleURL.scheme hasPrefix:@"http"]) {
         // Check if the current appVersion has been reported.
         NSString *appVersion = [[CodePushConfig current] appVersion];
-        if ([self isDeploymentStatusNotYetReported:appVersion]) {
-            [self recordDeploymentStatusReported:appVersion];
+        NSString *previousStatusReportIdentifier = [CodePushStatusReport getPreviousStatusReportIdentifier];
+        if (previousStatusReportIdentifier == nil) {
+            [CodePushStatusReport recordDeploymentStatusReported:appVersion];
             resolve(@{ @"appVersion": appVersion });
             return;
+        } else if (![previousStatusReportIdentifier isEqualToString:appVersion]) {
+            [CodePushStatusReport recordDeploymentStatusReported:appVersion];
+            if ([CodePushStatusReport isStatusReportIdentifierCodePushLabel:previousStatusReportIdentifier]) {
```

Comment:
I'm not sure I fully understand this code. If the `previousStatusReportIdentifier` isn't `nil`, when would it ever be set to an invalid report identifier? The code for this seems to simply check for `nil` (which you already do), and then check to see if it includes a colon. Do we need this check at all?